### DISCUSSION TITLE
Should the recommend zsh shellenv be stored to .zshenv instead of .zprofile?

### DISCUSSION THREAD

**Discussion by NReilingh** [Homebrew/discussions#1127] [Everyday usage]
I noticed while installing Homebrew on a new machine today that the next steps list:

```
- Add Homebrew to your PATH in /Users/nick/.zprofile:
    echo 'eval "$(/opt/homebrew/bin/brew shellenv)"' >> /Users/nick/.zprofile
    eval "$(/opt/homebrew/bin/brew shellenv)"
```

On my past machines, I would always have this in `.zshenv` instead of `.zprofile`. The differences are explained in this helpful article: https://scriptingosx.com/2019/06/moving-to-zsh-part-2-configuration-files/

Notably, `.zshenv` is sourced for ALL execution contexts, not just login shells. The general advice I've seen previously suggests that modifications to the `$PATH` are best made in `.zshenv`. Do people have strong opinions about this in the opposite direction, or should I try to move forward with a PR?

**Comment by carlocab** [comment]
I remember @richiksc mentioning that things don't quite work as expected if you add `/opt/homebrew/bin` to your `PATH` in `.zshenv`, but I forget the details and the explanation is buried in a large thread.

See also http://zsh.sourceforge.net/Guide/zshguide02.html#l9 for some pitfalls with setting your `PATH` in `.zshrc`.

**Reply by NReilingh** [reply]
Thanks! I haven't had any luck in finding previous discussions on this -- the only other discussion result mentioning `.zshenv` is https://github.com/Homebrew/discussions/discussions/448, and I don't get any results from the old Discourse.

The docs don't really list any pitfalls except that `~/.zshenv` isn't _quite_ as unskippable as `/etc/zshenv`, but the conditions in which it would be skipped would also apply to `.zprofile` and others.

**Reply by carlocab** [reply]
Nope, there's definitely more. See, for example, in 2.5.11:

>There's a slight difficulty in setting `$path` in `.zshenv` however

**Reply by richiksc** [reply]
Yes, I found that adding `/opt/homebrew/bin` to the path in `.zshenv` caused some issues. Turns out `.zshenv` is loaded before the system path is loaded on macOS (from `/etc/paths` through the `/usr/libexec/path_helper` script), so `/usr/bin` would be prepended to `/opt/homebrew/bin`, causing system binaries to take precedence over Homebrew-installed ones. This is obviously not a preferable behavior, especially if you want to use updated versions of system tools like `nano`, `git`, and `zsh` through Homebrew.

More info: https://scriptingosx.com/2017/05/where-paths-come-from/
Edit: I personally set my path in `.zshrc`.

**Reply by carlocab** [reply]
>Turns out `.zshenv` is loaded before the system path is loaded on macOS

Yep, that's a hurdle that's impossible to get around, I think. Thanks for the reminder, @richiksc!

**Reply by NReilingh** [reply]
Ah, I see -- I only read the 2.2 section about the startup files themselves. Makes sense that I wouldn't have encountered those issues with the PATH since I tend not to ever install packages that would clobber system binaries.

I'll consider this question answered. :-) Seems that different folks will have different reasons for wanting it in different places, so for everyone else just following the directions, `.zprofile` is fine.
